analysis/scala-implicit-browser/app.R

```diff
@@ -0,0 +1,350 @@
+library(shiny)
+library(magrittr)
+library(tidyverse)
+library(visNetwork)
+library(igraph)
+library(stringr)
+library(janitor)
+library(RMySQL)
+library(dbplyr)
+
+NODE_PARAM <- 1L
+NODE_FUN <- 2L
+
+load_data <- function() {
+  con <- DBI::dbConnect(RMySQL::MySQL(), dbname="scala", username="scala", password="scala", host="mysql")
+  projects <- tbl(con, "projects") %>% collect(n=Inf)
+  funs <- tbl(con, "funs")
+  params <- tbl(con, "params")
+  params_funs <- tbl(con, "params_funs")
+
+  list(
+    projects=projects,
+    funs=funs,
+    params=params,
+    params_funs=params_funs,
+    con=con
+  )
+}
+
+create_model <- function(proj, funs, params, params_funs) {
+  # convert the ids to simple indexes
+
+  edges <- data_frame(
+    from=match(params_funs$param, params$id),
+    to=match(params_funs$fun, funs$id) + nrow(params),
+    id=1:nrow(params_funs)
+  )
+
+  m_param_nodes <- params %>% transmute(
+    id=1:nrow(params),
+    node_type=NODE_PARAM,
+
+    name=name,
+    fqn=fqn,
+    type=type,
+    fqtn=str_replace(fqtn, "_root_.(.*)#", "\\1"),
+    kind=kind
+  )
+
+  v_param_nodes <- m_param_nodes %>% transmute(
+    id=1:nrow(params),
+    node_type=NODE_PARAM,
+
+    group="parameters",
+    shape="dot",
+    title=str_c("<pre><code>implicit ", kind, " ", fqn, ".", name, ": ", fqtn, "</code></pre>", sep=""),
+    color="red"
+  )
+
+  m_fun_nodes <- funs %>% transmute(
+    id=nrow(params)+(1:nrow(funs)),
+    node_type=NODE_FUN,
+
+    # FIXME: this is only good for github
+    url=str_c(str_c(proj$url, "blob", proj$version, path, sep="/"), "#L", line),
+    path=path,
+    line=line,
+    col=col,
+    module=str_replace(path, "(.*)/src/.*", "\\1"),
+    srcset=str_replace(path, ".*/src/([^/]+)/.*", "\\1"),
+    # TODO add code
+    code="", #str_replace_all(code, "\\\\n", "\n"),
+    # TODO: process symbol name
+    name=str_replace(name, fixed("_root_."), ""),
+    nargs=nargs
+  )
+
+
+  v_fun_nodes <- m_fun_nodes %>% transmute(
+    id=nrow(params)+(1:nrow(funs)),
+    node_type=NODE_FUN,
+
+    group="functions",
+    shape="dot",
+    title=str_c('<pre><a target="_blank" href="', url ,'">', path, "</a><br/><br/><code>", name, "</code></pre>"),
+    color="lightblue"
+  )
+
+  m_nodes <- bind_rows(m_param_nodes, m_fun_nodes)
+  v_nodes <- bind_rows(v_param_nodes, v_fun_nodes)
+
+  list(
+    nodes=m_nodes,
+    edges=edges,
+    graph=igraph::graph_from_data_frame(edges, vertices=v_nodes, directed=T)
+  )
+}
+
+ui <- fluidPage(
+   titlePanel("Implicit Browser"),
+
+   sidebarLayout(
+      sidebarPanel(
+        selectInput(
+          inputId="project",
+          label=h4("Project"),
+          choices=NULL
+        ),
+        h4("Filter"),
+        checkboxGroupInput(
+          inputId="filter_kind",
+          label=h5("by kind")
+        ),
+        checkboxGroupInput(
+          inputId="filter_srcset",
+          label=h5("by srcset")
```

Comment:
Perhaps a more descriptive name (just `source group`, or whatever. just full words), as a small detail.

Reply:
I will call it `source directory` - that seems to be sbt terminology.